### ISSUE TITLE
txn API can't register proxy services (and possibly other bugs)

### ISSUE THREAD

**Issue by banks** [hashicorp/consul#14341]
#### Overview of the Issue

The `/txn` API supports requesting multiple read and write operations across several data types in a single atomic batch. One such type is Service registrations.

In order to implement this, the HTTP endpoint translates the `api.*` struct types that the JSON body is parsed into into the internal `structs.*` types required for the internal RPC call. That translation is manually coded and has grown stale over the years, especially for the Service registration type which has evolved the most as we've implemented a service mesh.

Specifically, we have not updated the code below to copy several important new fields, at least: `Kind, Connect,  Proxy` (though there may be more). Which means you can't register Proxy services with `/txn` right now - it appears to work, but those fields are silently stripped resulting in a "normal" service type without that data.

https://github.com/hashicorp/consul/blob/13c04a13af70bcd8c00fcf55003212f584a92799/agent/txn_endpoint.go#L204-L226

#### Reproduction Steps

Steps to reproduce this issue:

1. Assuming a Consul server with connect enabled... try this:
   ```sh
   $ cat catalog-service-reg.json
   [
    {
      "Node": {
        "Verb": "set",
        "Node": {
          "Node": "fake-node",
          "Address": "127.0.0.1"
        }
      }
    },
    {
      "Service": {
        "Verb": "set",
        "Node": "fake-node",
        "Service": {
          "Service": "test",
          "Port": 4444
        }
      }
    },
    {
      "Service": {
        "Verb": "set",
        "Node": "fake-node",
        "Service": {
          "Service": "test-sidecar-proxy",
          "Port": 20000,
          "Kind": "connect-proxy",
          "Proxy": {
            "DestinationServiceName": "test",
            "DestinationServiceID": "test",
            "LocalServiceAddress": "127.0.0.1",
            "LocalServicePort": 4444
          }
        }
      }
    }
   ]
   $ curl -X PUT -d @catalog-service-reg.json "${CONSUL_HTTP_ADDR}/v1/txn"
   ```

 2. Observe the returned value (that was written) doesn't contain Kind or Proxy fields:
     ```
      {
        "Results": [
            {
                "Node": {
                    ...
                }
            },
            {
                "Service": {
                   ...
                }
            },
            {
                "Service": {
                    "ID": "test-sidecar-proxy",
                    "Service": "test-sidecar-proxy",
                    "Tags": null,
                    "Address": "",
                    "Meta": null,
                    "Port": 20000,
                    "Weights": {
                        "Passing": 1,
                        "Warning": 1
                    },
                    "EnableTagOverride": false,
                    "Proxy": {
                        "Mode": "",
                        "MeshGateway": {},
                        "Expose": {}
                    },
                    "Connect": {},
                    "Namespace": "default",
                    "CreateIndex": 58,
                    "ModifyIndex": 58
                }
            }
        ],
        "Errors": null
    }
     ```
This was on current `main` branch but has almost certainly been the case since those fields were added in 1.2 and later releases.

## Fix Considerations

 - [ ] Initially, audit the struct code linked above and add missing field translations. Note that there may well be others in other types like `Checks` etc. that have diverged over the years too, although `ServiceNode` has changed more than most types.
 - [ ] Consider a strategy to avoid missing this in future. 
    - The ideal might be to switch to machine-generating this translation like we do for Protobuf <-> struct with `mog`, however considering the impact here (no one has reported this issue in 3+ years) the effort required for that seems not very well justified for now. 
    - At least though we could add comments to a few well-known places especially in the `ServiceNode` struct that is most likely to change reminding future authors to update this code too - we have similar notices for other places where structs need to be kept in sync.

**Comment by asimrehman92** [comment]
@banks 
Hi Gopher, 
if this PR is still open, Could you please allow me to try to solve this issue?

**Comment by banks** [comment]
@asimrehman92 Thanks for your offer. It _looks_ like @huikang already fixed this in Consul 1.13 and later. Is that accurate Hui? Should we close this now?

Are you still experiencing this issue @asimrehman92?

**Comment by huikang** [comment]
This issue has been resolved since 1.13. Feel free to re-open if anything wrong is observed.

Thanks, @banks @asimrehman92
